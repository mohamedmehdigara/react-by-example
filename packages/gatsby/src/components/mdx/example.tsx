import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
