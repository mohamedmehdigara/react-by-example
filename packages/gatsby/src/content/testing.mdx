---
title: Testing
---

## Testing

- [Arrange-Act-Assert][aaa] pattern works great with React

[aaa]: https://duckduckgo.com/?q=arrange+act+assert

```yaml {example}
type: test
project: testing-dummy
module: src/index.test.ts
```

### `GifFinder`

- let's test a component that uses state & effects to find gifs

```yaml {example}
project: testing-components
style:
  width: 100%
  height: 50em
```

### Testing `GifFinder`

> The more your tests resemble the way your software is used,
> the more confidence they can give you.
>
> -- [`@testing-library`]

- [`create-react-app`] comes with [`@testing-library`]
- works with [angular], [vue], [plain DOM][dom], etc...
- no selectors, [forcing] you to write accessible components
- alternatively, use [enzyme]

[`create-react-app`]: https://create-react-app.dev/
[`@testing-library`]: https://testing-library.com/
[angular]: https://testing-library.com/docs/angular-testing-library/intro
[vue]: https://testing-library.com/docs/vue-testing-library/intro
[dom]: https://testing-library.com/docs/dom-testing-library/intro
[forcing]: https://coderefinery.nz/2020/10/21/forcing-functions-in-software-development/
[enzyme]: https://github.com/enzymejs/enzyme

```yaml {example}
type: test
project: testing-components-test
```
