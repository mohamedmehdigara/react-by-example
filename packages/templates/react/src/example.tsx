import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
