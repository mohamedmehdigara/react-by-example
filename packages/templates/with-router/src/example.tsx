import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
