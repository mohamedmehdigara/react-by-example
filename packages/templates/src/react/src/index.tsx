import React from 'react'
import ReactDOM from 'react-dom'
import example from './example'

ReactDOM.render(<>{example}</>, document.getElementById('root'))
