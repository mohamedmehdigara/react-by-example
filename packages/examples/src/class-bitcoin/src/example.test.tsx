import { act, cleanup, render, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { BitcoinAPI } from './bitcoin-api'
import { Bitcoin } from './example'
import { Deferred } from './testing'

jest.mock('./bitcoin-api')
const MockBitcoinAPI = BitcoinAPI as jest.Mocked<typeof BitcoinAPI>

describe('Bitcoin', () => {
  let component: RenderResult
  let getPriceDeferred: Deferred<number>

  beforeEach(() => {
    getPriceDeferred = new Deferred()
    MockBitcoinAPI.getPrice.mockReturnValue(getPriceDeferred.promise)
    component = render(<Bitcoin />)
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  describe('when price resolves', () => {
    beforeEach(async () => {
      getPriceDeferred.resolve(12345)
      await act(async () => {
        await getPriceDeferred.promise
      })
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    it('should show price', () => {
      expect(component.container).toHaveTextContent(/£12345/)
    })

    describe('when switching fiat', () => {
      beforeEach(async () => {
        getPriceDeferred = new Deferred()
        getPriceDeferred.resolve(23456)
        MockBitcoinAPI.getPrice.mockReturnValue(getPriceDeferred.promise)
        userEvent.click(component.getByText('$'))

        await act(async () => {
          await getPriceDeferred.promise
        })
      })

      it('should show price', () => {
        expect(component.container).toHaveTextContent(/\$23456/)
      })

      describe('when switching fiat', () => {
        beforeEach(async () => {
          getPriceDeferred = new Deferred()
          getPriceDeferred.resolve(345678)
          MockBitcoinAPI.getPrice.mockReturnValue(getPriceDeferred.promise)
          userEvent.click(component.getByText('£'))

          await act(async () => {
            await getPriceDeferred.promise
          })
        })

        it('should show price', () => {
          expect(component.container).toHaveTextContent(/£345678/)
        })
      })

      describe('when switching fiat quickly', () => {
        beforeEach(async () => {
          const previousGetPriceDeferred = new Deferred<number>()
          MockBitcoinAPI.getPrice.mockReturnValue(
            previousGetPriceDeferred.promise,
          )
          userEvent.click(component.getByText('£'))

          getPriceDeferred = new Deferred()
          getPriceDeferred.resolve(4567)
          MockBitcoinAPI.getPrice.mockReturnValue(getPriceDeferred.promise)
          userEvent.click(component.getByText('$'))

          previousGetPriceDeferred.resolve(5678)

          await act(async () => {
            await getPriceDeferred.promise
            await previousGetPriceDeferred.promise
          })
        })

        // TODO fix race-condition
        it.skip('should show price for latest fiat', () => {
          expect(component.container).toHaveTextContent(/\$4567/)
        })
      })
    })
  })

  describe('when request rejects', () => {
    beforeEach(async () => {
      getPriceDeferred.reject()
      await act(async () => {
        await getPriceDeferred.promise.catch(() => null)
      })
    })

    it('should snapshot', () => {
      expect(component.container).toMatchSnapshot()
    })

    it('should show error', () => {
      expect(component.container).toHaveTextContent(/£NaN/)
    })
  })
})
