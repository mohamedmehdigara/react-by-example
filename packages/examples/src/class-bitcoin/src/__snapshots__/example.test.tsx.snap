// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitcoin should snapshot 1`] = `
<div>
  <div>
    <button
      type="button"
    >
      £
    </button>
    <button
      type="button"
    >
      $
    </button>
    <br />
    <span>
      loading price...
    </span>
  </div>
</div>
`;

exports[`Bitcoin when price resolves should snapshot 1`] = `
<div>
  <div>
    <button
      type="button"
    >
      £
    </button>
    <button
      type="button"
    >
      $
    </button>
    <br />
    <span>
      £
      12345
    </span>
  </div>
</div>
`;

exports[`Bitcoin when request rejects should snapshot 1`] = `
<div>
  <div>
    <button
      type="button"
    >
      £
    </button>
    <button
      type="button"
    >
      $
    </button>
    <br />
    <span>
      £
      NaN
    </span>
  </div>
</div>
`;
